Add support for Flattened JWS

g2p opened the pull request:
This adds some new APIs for JWS in JSON form.

Currently, JSON structs carrying a single signature are supported (the "flattened" serialization); the general form could be added as necessary when users need multiple signatures over the same payload.

The APIs do not mirror the ones for compact JWS; I deliberately did not introduce an enum that could contain signed and unsigned, or serialized and unserialized forms.

g2p commented:
Note: #190 and #189 together allow a functional ACME client.

lawliet89 commented:
Thanks for the PR. I'll look over it ASAP. In the mean time, could you rebase/resolve the conflicts?

Greatly appreciate your contributions!

g2p commented:
Yup, rebased. Thanks for having a look!

lawliet89 commented on src/jws/flattened.rs:
```diff
@@ -0,0 +1,542 @@
+use serde::{de::DeserializeOwned, Deserialize, Deserializer, Serialize};
+
+// Not using CompactPart::to_bytes here, bounds are overly restrictive
+fn serialize_header<H: Serialize>(header: &Header<H>) -> Vec<u8> {
```
We should return a `Result` since the serialization can fail and not panic with unwraps.

g2p replied:
Done

lawliet89 commented on src/jws/flattened.rs:
```diff
@@ -0,0 +1,542 @@
+    r
+}
+
+/// Data that can be turned into a JWS
```
Please provide some examples on how this can be used. Also, please indicate that this is for flattened JWS.

lawliet89 commented on src/jws.rs:
```diff
@@ -2,1451 +2,8 @@
 //! you will want to look at the  [`Compact`](enum.Compact.html) enum.
-use std::str;
-use std::sync::Arc;
+mod compact;
+mod flattened;
```
Let's make these two modules `pub`. Please move the relevant compact JWS documentation to the module documentation of `compact` and we can leave a general documentation about JWS in this module.

We can leave the `pub use` below to maintain some compatibility.

lawliet89 commented on src/jws/flattened.rs:
```diff
@@ -0,0 +1,542 @@
+//! Flattened JWS signatures: see RFC 7515 section 7.2.2
```
Please provide some examples on how flattened JWS can be used with the library.

g2p replied:
Done

lawliet89 commented on src/jws/flattened.rs:
```diff
@@ -0,0 +1,542 @@
+    payload: Vec<u8>,
+}
+
+impl Signable {
```
Can we add a `sign` convenience function to call `SignedData::sign` please.

g2p replied:
Done

lawliet89 commented on src/jws/flattened.rs:
```diff
@@ -0,0 +1,542 @@
+            signatures: (),
+            unprotected_header: (),
+        };
+        serde_json::to_string(&s).unwrap()
```
Let's avoid unwrapping and return `Result`.

g2p replied:
I added a comment explaining why this shouldn't fail instead.

lawliet89 commented:
This might require a few rounds of back and forth because I am not entirely familiar with JWS flattened at this point.

g2p commented:
I did a round of updates addressing feedback.

About exposing the module structure, I'd argue the current module split should not be exposed.
It reflects the way the APIs have been added, the current way does not integrate things as much as it could so as to minimize disruption.
The implementation is separate because I could not integrate with the way Compact is implemented. I left the Compact enum alone, and reused some lower-level stuff, meaning compact.rs contains things (like Header, RegisteredHeader, Secret) that are now shared.

Moving things further, that depends on your wishes:
- Are you open to SignedData being actually something shared as well, with a way to serialize into a compact form? Or should the code paths remain mostly separate?
- Are you open to anything that deprecates or breaks compatibility?

lawliet89 commented:
> Are you open to SignedData being actually something shared as well, with a way to serialize into a compact form? Or should the code paths remain mostly separate?

Wouldn't mind if we can reuse code as much as possible. But it boils down to whether it makes things too complex to implement or not.

> Are you open to anything that deprecates or breaks compatibility?

Possibly. Your previous PR (https://github.com/lawliet89/biscuit/pull/189) has already introduced a breaking change, albeit a small one. I would like to mimise this if possible.

lawliet89 commented on src/jws/flattened.rs:
```diff
@@ -133,6 +141,8 @@ impl SignedData {
         };
+        // This shouldn't fail, because FlattenedRaw strucs are
+        // always representable in JSON
         serde_json::to_string(&s).unwrap()
```
Perhaps use an `expect` to give some meaningful error messages if it ever panics.

g2p replied:
Done

lawliet89 commented on src/jws/flattened.rs:
```diff
@@ -0,0 +1,637 @@
+    serde_json::to_vec(header)
+}
+
+/// Warning: pay attention to parameter order
```
```suggestion
// Warning: pay attention to parameter order
```

g2p replied:
Applied

lawliet89 reviewed (commented):
I think I'm OK with what you have changed so far. I'm ready to merge it in unless you want to make more changes.

g2p commented:
> I think I'm OK with what you have changed so far. I'm ready to merge it in unless you want to make more changes.

Yup, I'm happy with it as well!

I've added a commit to move the stuff that turns out to be shared back into jws.rs.
I don't think anything else really needs to be done; anything further (like sharing the implementations more or adding functionality to SignedData) can be a different PR.

g2p commented:
Hello,
could you release the crate so I can make use of this?
Thank you

lawliet89 commented:
Is it OK if I release it as a beta first?

g2p commented:
Yes, that's fine

lawliet89 commented:
See https://crates.io/crates/biscuit/0.6.0-beta1
